// @aws-amplify/backend-cli/lib/commands/sandbox/sandbox-secret/sandbox_secret_set_command.js
import { AmplifyPrompter } from '@aws-amplify/cli-core';
/**
 * Command to set sandbox secret.
 */
export class SandboxSecretSetCommand {
    sandboxIdResolver;
    secretClient;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Set sandbox secret command.
     */
    constructor(sandboxIdResolver, secretClient) {
        this.sandboxIdResolver = sandboxIdResolver;
        this.secretClient = secretClient;
        this.command = 'set <secret-name>';
        this.describe = 'Set a sandbox secret';
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        const secretVal = await AmplifyPrompter.secretValue();
        await this.secretClient.setSecret(await this.sandboxIdResolver.resolve(args.identifier), args.secretName, secretVal);
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs.positional('secret-name', {
            describe: 'Name of the secret to set',
            type: 'string',
            demandOption: true,
        });
    };
}
