// @aws-amplify/backend-cli/lib/commands/sandbox/sandbox-secret/sandbox_secret_remove_command.js
/**
 * Command to remove sandbox secret.
 */
export class SandboxSecretRemoveCommand {
    sandboxIdResolver;
    secretClient;
    /**
     * @inheritDoc
     */
    command;
    /**
     * @inheritDoc
     */
    describe;
    /**
     * Remove sandbox secret command.
     */
    constructor(sandboxIdResolver, secretClient) {
        this.sandboxIdResolver = sandboxIdResolver;
        this.secretClient = secretClient;
        this.command = 'remove <secret-name>';
        this.describe = 'Remove a sandbox secret';
    }
    /**
     * @inheritDoc
     */
    handler = async (args) => {
        const sandboxBackendIdentifier = await this.sandboxIdResolver.resolve(args.identifier);
        await this.secretClient.removeSecret(sandboxBackendIdentifier, args.secretName);
    };
    /**
     * @inheritDoc
     */
    builder = (yargs) => {
        return yargs.positional('secret-name', {
            describe: 'Name of the secret to remove',
            type: 'string',
            demandOption: true,
        });
    };
}
