// @aws-amplify/sandbox/lib/sandbox_executor.js
import debounce from 'debounce-promise';
import { SecretError } from '@aws-amplify/backend-secret';
import { LogLevel } from '@aws-amplify/cli-core';
import { AmplifyFault, AmplifyUserError } from '@aws-amplify/platform-core';
import { SSMServiceException } from '@aws-sdk/client-ssm';
/**
 * Execute CDK commands.
 */
export class AmplifySandboxExecutor {
    backendDeployer;
    secretClient;
    printer;
    /**
     * Function that invokes the callback with debounce.
     * Debounce is needed in case multiple duplicate events are received.
     */
    invoke = debounce(async (callback) => await callback(), 100);
    /**
     * Creates an AmplifySandboxExecutor instance
     */
    constructor(backendDeployer, secretClient, printer) {
        this.backendDeployer = backendDeployer;
        this.secretClient = secretClient;
        this.printer = printer;
    }
    /**
     * Deploys sandbox
     */
    deploy = async (backendId, validateAppSourcesProvider) => {
        this.printer.log('[Sandbox] Executing command `deploy`', LogLevel.DEBUG);
        const secretLastUpdated = await this.getSecretLastUpdated(backendId);
        return this.invoke(() => {
            // it's important to get information here so that information
            // doesn't get lost while debouncing
            const validateAppSources = validateAppSourcesProvider();
            return this.backendDeployer.deploy(backendId, {
                secretLastUpdated,
                validateAppSources,
            });
        });
    };
    /**
     * Destroy sandbox. Do not swallow errors
     */
    destroy = (backendId) => {
        this.printer.log('[Sandbox] Executing command `destroy`', LogLevel.DEBUG);
        return this.invoke(() => this.backendDeployer.destroy(backendId));
    };
    getSecretLastUpdated = async (backendId) => {
        let secrets = undefined;
        try {
            secrets = await this.secretClient.listSecrets(backendId);
        }
        catch (err) {
            if (err instanceof SecretError &&
                err.cause &&
                err.cause instanceof SSMServiceException) {
                if (err.cause.name === 'ExpiredTokenException') {
                    throw new AmplifyUserError('SecretsExpiredTokenError', {
                        message: 'Fetching the list of secrets failed due to expired tokens',
                        resolution: 'Please refresh your credentials and try again',
                    }, err);
                }
            }
            throw new AmplifyFault('ListSecretsFailedFault', {
                message: 'Fetching the list of secrets failed',
            }, err);
        }
        let latestTimestamp = -1;
        let secretLastUpdate;
        secrets.forEach((secret) => {
            if (!secret.lastUpdated) {
                return;
            }
            const curTimeStamp = secret.lastUpdated.getTime();
            if (curTimeStamp > 0 && curTimeStamp > latestTimestamp) {
                latestTimestamp = curTimeStamp;
                secretLastUpdate = secret.lastUpdated;
            }
        });
        return secretLastUpdate;
    };
}
